gRPC: use of `userProject` options results in error from gcs when used with credentials which define quota_project_id

Affected version(s): 2.14.0+

#### Steps to reproduce

1. Set project via gcloud config set project $projectId
2. Create adc following instructions from https://google.aip.dev/auth/4113
3. <details>
   <summary>JUnit tests to repro failure</summary>

   ```java
    import static com.google.common.truth.Truth.assertThat;
    
    import com.google.cloud.storage.Bucket;
    import com.google.cloud.storage.BucketFixture;
    import com.google.cloud.storage.BucketInfo;
    import com.google.cloud.storage.Storage;
    import com.google.cloud.storage.Storage.BucketField;
    import com.google.cloud.storage.Storage.BucketGetOption;
    import com.google.cloud.storage.Storage.BucketSourceOption;
    import com.google.cloud.storage.Storage.BucketTargetOption;
    import com.google.cloud.storage.StorageFixture;
    import com.google.cloud.storage.conformance.retry.CleanupStrategy;
    import org.junit.AfterClass;
    import org.junit.BeforeClass;
    import org.junit.ClassRule;
    import org.junit.Test;
    import org.slf4j.Logger;
    import org.slf4j.LoggerFactory;
    import org.slf4j.Marker;
    import org.slf4j.MarkerFactory;
    import org.slf4j.bridge.SLF4JBridgeHandler;
    
    public final class ITRequesterPaysTest {
      static {
        SLF4JBridgeHandler.removeHandlersForRootLogger();
        SLF4JBridgeHandler.install();
      }
      private static final Logger LOGGER = LoggerFactory.getLogger(ITRequesterPaysTest.class);
      private static final Marker TRACE_ENTER = MarkerFactory.getMarker("enter");
      private static final Marker TRACE_EXIT = MarkerFactory.getMarker("exit");
    
      @ClassRule
      public static final StorageFixture sfGadc = StorageFixture.defaultGrpc();
    
      private static final BucketFixture bf =
          BucketFixture.newBuilder()
              .setCleanupStrategy(CleanupStrategy.NEVER)
              .setBucketNameFmtString("java-storage-grpc-%s")
              .setHandle(sfJadc::getInstance)
              .build();
    
      private static final String bucketName = bf.newBucketName();
    
      @BeforeClass
      public static void beforeClass() {
        LOGGER.trace(TRACE_ENTER, "beforeClass()");
        try {
          BucketInfo info1 = BucketInfo.newBuilder(bucketName).setRequesterPays(true).build();
          Bucket bucket = sfJadc.getInstance().create(info1, BucketTargetOption.projection("noAcl"));
          assertThat(bucket.requesterPays()).isTrue();
        } finally {
          LOGGER.trace(TRACE_EXIT, "beforeClass()");
        }
      }
    
      @AfterClass
      public static void afterClass() {
        LOGGER.trace(TRACE_ENTER, "afterClass()");
        try {
          String projectId = sfJadc.getInstance().getOptions().getProjectId();
          sfJadc.getInstance().delete(bucketName, BucketSourceOption.userProject(projectId));
        } finally {
          LOGGER.trace(TRACE_EXIT, "afterClass()");
        }
      }
    
      @Test
      public void grpc_adc() {
        LOGGER.trace(TRACE_ENTER, "grpc_adc()");
        try {
          doTest(sfGadc.getInstance());
        } finally {
          LOGGER.trace(TRACE_EXIT, "grpc_adc()");
        }
      }
    
      private static void doTest(Storage storage) {
        LOGGER.trace(TRACE_ENTER, "doTest(storage : {})", storage);
        try {
          String projectId = storage.getOptions().getProjectId();
          doGet(storage, projectId);
          // doList(storage, projectId);
        } finally {
          LOGGER.trace(TRACE_EXIT, "doTest(storage : {})", storage);
        }
      }
    
      private static void doGet(Storage storage, String projectId) {
        Bucket bucket =
            storage
                .get(
                    bucketName,
                    BucketGetOption.fields(BucketField.ID, BucketField.BILLING),
                    BucketGetOption.userProject(projectId));
        assertThat(bucket.requesterPays()).isTrue();
      }
    }

   ```
   </details>

#### Stack trace
```
com.google.cloud.storage.StorageException: INVALID_ARGUMENT: User project specified in the request is invalid.

	at com.google.cloud.storage.StorageException.asStorageException(StorageException.java:144)
	at com.google.cloud.storage.StorageException.coalesce(StorageException.java:117)
	at com.google.cloud.storage.Retrying.run(Retrying.java:96)
	at com.google.cloud.storage.GrpcStorageImpl.get(GrpcStorageImpl.java:376)
	at com.google.cloud.storage.it.ITObjectTest.unsetRequesterPays(ITObjectTest.java:154)
	at com.google.cloud.storage.it.ITObjectTest.testListBlobRequesterPays(ITObjectTest.java:504)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.google.api.gax.rpc.InvalidArgumentException: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: User project specified in the request is invalid.
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:92)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:98)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:66)
	at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
	at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:67)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1132)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:31)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1270)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:1038)
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:808)
	at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:574)
	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:544)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at com.google.api.gax.grpc.ChannelPool$ReleasingClientCall$1.onClose(ChannelPool.java:535)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:563)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:70)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:744)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:723)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	... 1 more
	Suppressed: com.google.api.gax.rpc.AsyncTaskException: Asynchronous task failed
		at com.google.api.gax.rpc.ApiExceptions.callAndTranslateApiException(ApiExceptions.java:57)
		at com.google.api.gax.rpc.UnaryCallable.call(UnaryCallable.java:100)
		at com.google.cloud.storage.GrpcStorageImpl.lambda$get$2(GrpcStorageImpl.java:379)
		at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:103)
		at com.google.cloud.RetryHelper.run(RetryHelper.java:76)
		at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
		at com.google.cloud.storage.Retrying.run(Retrying.java:93)
		at com.google.cloud.storage.GrpcStorageImpl.get(GrpcStorageImpl.java:376)
		at com.google.cloud.storage.it.ITObjectTest.unsetRequesterPays(ITObjectTest.java:154)
		at com.google.cloud.storage.it.ITObjectTest.testListBlobRequesterPays(ITObjectTest.java:504)
		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:498)
		at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
		at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
		at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
		at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
		at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
		at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
		at java.util.concurrent.FutureTask.run(FutureTask.java:266)
		... 1 more
Caused by: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: User project specified in the request is invalid.
	at io.grpc.Status.asRuntimeException(Status.java:535)
	... 14 more
```

#### Current Workaround

Decorate default credentials to hide quota_project_id
```java
Storage s = StorageOptions.grpc()
  .setCredentials(new QuotaProjectIdHidingCredentials(GoogleCredentials.getApplicationDefault()))
  .build()
  .getService();
```

Internal reference bug: b/248605515

--- a/google-cloud-storage/src/main/java/com/google/cloud/storage/GrpcStorageImpl.java
+++ b/google-cloud-storage/src/main/java/com/google/cloud/storage/GrpcStorageImpl.java
@@ -67,6 +67,7 @@ import com.google.cloud.storage.UnifiedOpts.ObjectSourceOpt;
 import com.google.cloud.storage.UnifiedOpts.ObjectTargetOpt;
 import com.google.cloud.storage.UnifiedOpts.Opts;
 import com.google.cloud.storage.UnifiedOpts.ProjectId;
+import com.google.cloud.storage.UnifiedOpts.UserProject;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
@@ -159,11 +160,15 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
   final GrpcRetryAlgorithmManager retryAlgorithmManager;
   final SyntaxDecoders syntaxDecoders;
 
+  // workaround for https://github.com/googleapis/java-storage/issues/1736
+  private final Opts<UserProject> defaultOpts;
   @Deprecated private final ProjectId defaultProjectId;
 
-  GrpcStorageImpl(GrpcStorageOptions options, StorageClient storageClient) {
+  GrpcStorageImpl(
+      GrpcStorageOptions options, StorageClient storageClient, Opts<UserProject> defaultOpts) {
     super(options);
     this.storageClient = storageClient;
+    this.defaultOpts = defaultOpts;
     this.codecs = Conversions.grpc();
     this.retryAlgorithmManager = options.getRetryAlgorithmManager();
     this.syntaxDecoders = new SyntaxDecoders();
@@ -182,7 +187,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   @Override
   public Bucket create(BucketInfo bucketInfo, BucketTargetOption... options) {
-    Opts<BucketTargetOpt> opts = Opts.unwrap(options).resolveFrom(bucketInfo);
+    Opts<BucketTargetOpt> opts = Opts.unwrap(options).resolveFrom(bucketInfo).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     com.google.storage.v2.Bucket bucket = codecs.bucketInfo().encode(bucketInfo);
@@ -215,7 +220,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
       BlobInfo blobInfo, byte[] content, int offset, int length, BlobTargetOption... options) {
     requireNonNull(blobInfo, "blobInfo must be non null");
     requireNonNull(content, "content must be non null");
-    Opts<ObjectTargetOpt> opts = Opts.unwrap(options).resolveFrom(blobInfo);
+    Opts<ObjectTargetOpt> opts = Opts.unwrap(options).resolveFrom(blobInfo).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     WriteObjectRequest req = getWriteObjectRequest(blobInfo, opts);
@@ -267,7 +272,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
       throw new StorageException(0, path + " is a directory");
     }
 
-    Opts<ObjectTargetOpt> opts = Opts.unwrap(options).resolveFrom(blobInfo);
+    Opts<ObjectTargetOpt> opts = Opts.unwrap(options).resolveFrom(blobInfo).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     WriteObjectRequest req = getWriteObjectRequest(blobInfo, opts);
@@ -335,7 +340,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
       throws IOException {
     requireNonNull(blobInfo, "blobInfo must be non null");
 
-    Opts<ObjectTargetOpt> opts = Opts.unwrap(options).resolveFrom(blobInfo);
+    Opts<ObjectTargetOpt> opts = Opts.unwrap(options).resolveFrom(blobInfo).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     WriteObjectRequest req = getWriteObjectRequest(blobInfo, opts);
@@ -369,7 +374,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   @Override
   public Bucket get(String bucket, BucketGetOption... options) {
-    Opts<BucketSourceOpt> opts = Opts.unwrap(options);
+    Opts<BucketSourceOpt> opts = Opts.unwrap(options).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     GetBucketRequest.Builder builder =
@@ -384,7 +389,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   @Override
   public Bucket lockRetentionPolicy(BucketInfo bucket, BucketTargetOption... options) {
-    Opts<BucketTargetOpt> opts = Opts.unwrap(options).resolveFrom(bucket);
+    Opts<BucketTargetOpt> opts = Opts.unwrap(options).resolveFrom(bucket).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     LockBucketRetentionPolicyRequest.Builder builder =
@@ -406,7 +411,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   @Override
   public Blob get(BlobId blob, BlobGetOption... options) {
-    Opts<ObjectSourceOpt> opts = Opts.unwrap(options).resolveFrom(blob);
+    Opts<ObjectSourceOpt> opts = Opts.unwrap(options).resolveFrom(blob).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     GetObjectRequest.Builder builder =
@@ -435,7 +440,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   @Override
   public Page<Bucket> list(BucketListOption... options) {
-    Opts<BucketListOpt> opts = Opts.unwrap(options);
+    Opts<BucketListOpt> opts = Opts.unwrap(options).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     ListBucketsRequest request =
@@ -462,7 +467,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   @Override
   public Page<Blob> list(String bucket, BlobListOption... options) {
-    Opts<ObjectListOpt> opts = Opts.unwrap(options);
+    Opts<ObjectListOpt> opts = Opts.unwrap(options).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     ListObjectsRequest.Builder builder =
@@ -481,7 +486,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   @Override
   public Bucket update(BucketInfo bucketInfo, BucketTargetOption... options) {
-    Opts<BucketTargetOpt> opts = Opts.unwrap(options).resolveFrom(bucketInfo);
+    Opts<BucketTargetOpt> opts = Opts.unwrap(options).resolveFrom(bucketInfo).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     com.google.storage.v2.Bucket bucket = codecs.bucketInfo().encode(bucketInfo);
@@ -503,7 +508,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   @Override
   public Blob update(BlobInfo blobInfo, BlobTargetOption... options) {
-    Opts<ObjectTargetOpt> opts = Opts.unwrap(options).resolveFrom(blobInfo);
+    Opts<ObjectTargetOpt> opts = Opts.unwrap(options).resolveFrom(blobInfo).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     Object object = codecs.blobInfo().encode(blobInfo);
@@ -530,7 +535,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   @Override
   public boolean delete(String bucket, BucketSourceOption... options) {
-    Opts<BucketSourceOpt> opts = Opts.unwrap(options);
+    Opts<BucketSourceOpt> opts = Opts.unwrap(options).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     DeleteBucketRequest.Builder builder =
@@ -555,7 +560,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   @Override
   public boolean delete(BlobId blob, BlobSourceOption... options) {
-    Opts<ObjectSourceOpt> opts = Opts.unwrap(options).resolveFrom(blob);
+    Opts<ObjectSourceOpt> opts = Opts.unwrap(options).resolveFrom(blob).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     DeleteObjectRequest.Builder builder =
@@ -587,7 +592,9 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
   @Override
   public Blob compose(ComposeRequest composeRequest) {
     Opts<ObjectTargetOpt> opts =
-        Opts.unwrap(composeRequest.getTargetOptions()).resolveFrom(composeRequest.getTarget());
+        Opts.unwrap(composeRequest.getTargetOptions())
+            .resolveFrom(composeRequest.getTarget())
+            .prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     ComposeObjectRequest.Builder builder = ComposeObjectRequest.newBuilder();
@@ -609,8 +616,12 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
     BlobId src = copyRequest.getSource();
     BlobInfo dst = copyRequest.getTarget();
     Opts<ObjectSourceOpt> srcOpts =
-        Opts.unwrap(copyRequest.getSourceOptions()).projectAsSource().resolveFrom(src);
-    Opts<ObjectTargetOpt> dstOpts = Opts.unwrap(copyRequest.getTargetOptions()).resolveFrom(dst);
+        Opts.unwrap(copyRequest.getSourceOptions())
+            .projectAsSource()
+            .resolveFrom(src)
+            .prepend(defaultOpts);
+    Opts<ObjectTargetOpt> dstOpts =
+        Opts.unwrap(copyRequest.getTargetOptions()).resolveFrom(dst).prepend(defaultOpts);
 
     Mapper<RewriteObjectRequest.Builder> mapper =
         srcOpts.rewriteObjectsRequest().andThen(dstOpts.rewriteObjectsRequest());
@@ -684,7 +695,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   @Override
   public GrpcBlobReadChannel reader(BlobId blob, BlobSourceOption... options) {
-    Opts<ObjectSourceOpt> opts = Opts.unwrap(options).resolveFrom(blob);
+    Opts<ObjectSourceOpt> opts = Opts.unwrap(options).resolveFrom(blob).prepend(defaultOpts);
     ReadObjectRequest request = getReadObjectRequest(blob, opts);
     Set<StatusCode.Code> codes = resultRetryAlgorithmToCodes(retryAlgorithmManager.getFor(request));
     GrpcCallContext grpcCallContext = GrpcCallContext.createDefault().withRetryableCodes(codes);
@@ -722,7 +733,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   @Override
   public GrpcBlobWriteChannel writer(BlobInfo blobInfo, BlobWriteOption... options) {
-    Opts<ObjectTargetOpt> opts = Opts.unwrap(options).resolveFrom(blobInfo);
+    Opts<ObjectTargetOpt> opts = Opts.unwrap(options).resolveFrom(blobInfo).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     WriteObjectRequest req = getWriteObjectRequest(blobInfo, opts);
@@ -844,7 +855,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
   @Override
   public Acl getAcl(String bucket, Entity entity, BucketSourceOption... options) {
     try {
-      Opts<BucketSourceOpt> opts = Opts.unwrap(options);
+      Opts<BucketSourceOpt> opts = Opts.unwrap(options).prepend(defaultOpts);
       com.google.storage.v2.Bucket resp = getBucketWithAcls(bucket, opts);
 
       Predicate<BucketAccessControl> entityPredicate =
@@ -874,7 +885,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
   @Override
   public boolean deleteAcl(String bucket, Entity entity, BucketSourceOption... options) {
     try {
-      Opts<BucketSourceOpt> opts = Opts.unwrap(options);
+      Opts<BucketSourceOpt> opts = Opts.unwrap(options).prepend(defaultOpts);
       com.google.storage.v2.Bucket resp = getBucketWithAcls(bucket, opts);
       String encode = codecs.entity().encode(entity);
 
@@ -928,7 +939,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
   @Override
   public Acl updateAcl(String bucket, Acl acl, BucketSourceOption... options) {
     try {
-      Opts<BucketSourceOpt> opts = Opts.unwrap(options);
+      Opts<BucketSourceOpt> opts = Opts.unwrap(options).prepend(defaultOpts);
       com.google.storage.v2.Bucket resp = getBucketWithAcls(bucket, opts);
       BucketAccessControl encode = codecs.bucketAcl().encode(acl);
       String entity = encode.getEntity();
@@ -966,7 +977,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
   @Override
   public List<Acl> listAcls(String bucket, BucketSourceOption... options) {
     try {
-      Opts<BucketSourceOpt> opts = Opts.unwrap(options);
+      Opts<BucketSourceOpt> opts = Opts.unwrap(options).prepend(defaultOpts);
       com.google.storage.v2.Bucket resp = getBucketWithAcls(bucket, opts);
       return resp.getAclList().stream()
           .map(codecs.bucketAcl()::decode)
@@ -1211,7 +1222,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   @Override
   public HmacKey createHmacKey(ServiceAccount serviceAccount, CreateHmacKeyOption... options) {
-    Opts<HmacKeyTargetOpt> opts = Opts.unwrap(options);
+    Opts<HmacKeyTargetOpt> opts = Opts.unwrap(options).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     CreateHmacKeyRequest request =
@@ -1236,7 +1247,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   @Override
   public Page<HmacKeyMetadata> listHmacKeys(ListHmacKeysOption... options) {
-    Opts<HmacKeyListOpt> opts = Opts.unwrap(options);
+    Opts<HmacKeyListOpt> opts = Opts.unwrap(options).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
 
@@ -1264,7 +1275,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   @Override
   public HmacKeyMetadata getHmacKey(String accessId, GetHmacKeyOption... options) {
-    Opts<HmacKeySourceOpt> opts = Opts.unwrap(options);
+    Opts<HmacKeySourceOpt> opts = Opts.unwrap(options).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     GetHmacKeyRequest request =
@@ -1283,7 +1294,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   @Override
   public void deleteHmacKey(HmacKeyMetadata hmacKeyMetadata, DeleteHmacKeyOption... options) {
-    Opts<HmacKeyTargetOpt> opts = Opts.unwrap(options);
+    Opts<HmacKeyTargetOpt> opts = Opts.unwrap(options).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     DeleteHmacKeyRequest req =
@@ -1304,7 +1315,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
   @Override
   public HmacKeyMetadata updateHmacKeyState(
       HmacKeyMetadata hmacKeyMetadata, HmacKeyState state, UpdateHmacKeyOption... options) {
-    Opts<HmacKeyTargetOpt> opts = Opts.unwrap(options);
+    Opts<HmacKeyTargetOpt> opts = Opts.unwrap(options).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     com.google.storage.v2.HmacKeyMetadata encode =
@@ -1323,7 +1334,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   @Override
   public Policy getIamPolicy(String bucket, BucketSourceOption... options) {
-    Opts<BucketSourceOpt> opts = Opts.unwrap(options);
+    Opts<BucketSourceOpt> opts = Opts.unwrap(options).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     GetIamPolicyRequest.Builder builder =
@@ -1338,7 +1349,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   @Override
   public Policy setIamPolicy(String bucket, Policy policy, BucketSourceOption... options) {
-    Opts<BucketSourceOpt> opts = Opts.unwrap(options);
+    Opts<BucketSourceOpt> opts = Opts.unwrap(options).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     SetIamPolicyRequest req =
@@ -1356,7 +1367,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
   @Override
   public List<Boolean> testIamPermissions(
       String bucket, List<String> permissions, BucketSourceOption... options) {
-    Opts<BucketSourceOpt> opts = Opts.unwrap(options);
+    Opts<BucketSourceOpt> opts = Opts.unwrap(options).prepend(defaultOpts);
     GrpcCallContext grpcCallContext =
         opts.grpcMetadataMapper().apply(GrpcCallContext.createDefault());
     TestIamPermissionsRequest req =
@@ -1706,7 +1717,7 @@ final class GrpcStorageImpl extends BaseService<StorageOptions> implements Stora
 
   private UnbufferedReadableByteChannelSession<Object> unbufferedReadSession(
       BlobId blob, BlobSourceOption[] options) {
-    Opts<ObjectSourceOpt> opts = Opts.unwrap(options).resolveFrom(blob);
+    Opts<ObjectSourceOpt> opts = Opts.unwrap(options).resolveFrom(blob).prepend(defaultOpts);
     ReadObjectRequest readObjectRequest = getReadObjectRequest(blob, opts);
     Set<StatusCode.Code> codes =
         resultRetryAlgorithmToCodes(retryAlgorithmManager.getFor(readObjectRequest));

--- a/google-cloud-storage/src/main/java/com/google/cloud/storage/GrpcStorageOptions.java
+++ b/google-cloud-storage/src/main/java/com/google/cloud/storage/GrpcStorageOptions.java
@@ -31,15 +31,19 @@ import com.google.api.gax.retrying.StreamResumptionStrategy;
 import com.google.api.gax.rpc.HeaderProvider;
 import com.google.api.gax.rpc.NoHeaderProvider;
 import com.google.api.gax.rpc.StatusCode.Code;
+import com.google.api.gax.rpc.internal.QuotaProjectIdHidingCredentials;
 import com.google.auth.Credentials;
 import com.google.cloud.NoCredentials;
 import com.google.cloud.ServiceFactory;
 import com.google.cloud.ServiceOptions;
 import com.google.cloud.ServiceRpc;
 import com.google.cloud.TransportOptions;
+import com.google.cloud.Tuple;
 import com.google.cloud.grpc.GrpcTransportOptions;
 import com.google.cloud.spi.ServiceRpcFactory;
 import com.google.cloud.storage.TransportCompatibility.Transport;
+import com.google.cloud.storage.UnifiedOpts.Opts;
+import com.google.cloud.storage.UnifiedOpts.UserProject;
 import com.google.cloud.storage.spi.StorageRpcFactory;
 import com.google.common.base.MoreObjects;
 import com.google.common.collect.ImmutableSet;
@@ -50,6 +54,10 @@ import com.google.storage.v2.StorageSettings;
 import io.grpc.ManagedChannelBuilder;
 import java.io.IOException;
 import java.net.URI;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Map.Entry;
 import java.util.Set;
 import org.checkerframework.checker.nullness.qual.NonNull;
 import org.checkerframework.checker.nullness.qual.Nullable;
@@ -99,6 +107,41 @@ public final class GrpcStorageOptions extends StorageOptions
 
   @InternalApi
   StorageSettings getStorageSettings() throws IOException {
+    return resolveSettingsAndOpts().x();
+  }
+
+  /**
+   * We have to perform several introspections and detections to cross-wire/support several features
+   * that are either gapic primitives, ServieOption primitives or GCS semantic requirements.
+   *
+   * <h2>Requester Pays, {@code quota_project_id} and {@code userProject}</h2>
+   *
+   * When using the JSON Api operations destined for requester pays buckets can identify the project
+   * for billing and quota attribution by specifying either {@code userProject} query parameter or
+   * {@code x-goog-user-project} HTTP Header.
+   *
+   * <p>If the credentials being used contain the property {@code quota_project_id} this value will
+   * automatically be set to the {@code x-goog-user-project} header for both JSON and GAPIC. In the
+   * case of JSON this isn't an issue, as any {@code userProject} query parameter takes precedence.
+   * However, in gRPC/GAPIC there isn't a {@code userProject} query parameter, instead we are adding
+   * {@code x-goog-user-project} to the request context as metadata. If the credentials set the
+   * request metadata and we set the request metadata it results in two different entries in the
+   * request. This creates ambiguity for GCS which then rejects the request.
+   *
+   * <p>To account for this and to provide a similarly level of precedence we are introspecting the
+   * credentials and service options to save any {@code quota_project_id} into an {@link
+   * UserProject} which is then used by {@link GrpcStorageImpl} to resolve individual request
+   * metadata.
+   *
+   * <h3>The precedence we want to provide is as follows</h3>
+   *
+   * <ol>
+   *   <li>Any "userProject" Option provided to an individual method
+   *   <li>Any Non-empty value for {@link #getQuotaProjectId()}
+   *   <li>Any {@code x-goog-user-project} provided by {@link #credentials}
+   * </ol>
+   */
+  private Tuple<StorageSettings, Opts<UserProject>> resolveSettingsAndOpts() throws IOException {
     String endpoint = getHost();
     URI uri = URI.create(endpoint);
     String scheme = uri.getScheme();
@@ -114,11 +157,33 @@ public final class GrpcStorageOptions extends StorageOptions
         break;
     }
 
+    Opts<UserProject> defaultOpts = Opts.empty();
     CredentialsProvider credentialsProvider;
     if (credentials instanceof NoCredentials) {
       credentialsProvider = NoCredentialsProvider.create();
     } else {
-      credentialsProvider = FixedCredentialsProvider.create(credentials);
+      boolean foundQuotaProject = false;
+      if (credentials.hasRequestMetadata()) {
+        Map<String, List<String>> requestMetadata = credentials.getRequestMetadata(uri);
+        for (Entry<String, List<String>> e : requestMetadata.entrySet()) {
+          String key = e.getKey();
+          if ("x-goog-user-project".equals(key.trim().toLowerCase(Locale.ENGLISH))) {
+            List<String> value = e.getValue();
+            if (!value.isEmpty()) {
+              foundQuotaProject = true;
+              defaultOpts = Opts.from(UnifiedOpts.userProject(value.get(0)));
+              break;
+            }
+          }
+        }
+      }
+      if (foundQuotaProject) {
+        // fix for https://github.com/googleapis/java-storage/issues/1736
+        credentialsProvider =
+            FixedCredentialsProvider.create(new QuotaProjectIdHidingCredentials(credentials));
+      } else {
+        credentialsProvider = FixedCredentialsProvider.create(credentials);
+      }
     }
 
     HeaderProvider internalHeaderProvider =
@@ -134,6 +199,12 @@ public final class GrpcStorageOptions extends StorageOptions
             .setEndpoint(endpoint)
             .setCredentialsProvider(credentialsProvider)
             .setClock(getClock());
+
+    // this MUST come after credentials, service options set value has higher priority than creds
+    String quotaProjectId = this.getQuotaProjectId();
+    if (quotaProjectId != null && !quotaProjectId.isEmpty()) {
+      defaultOpts = Opts.from(UnifiedOpts.userProject(quotaProjectId));
+    }
 
     builder.setHeaderProvider(this.getMergedHeaderProvider(new NoHeaderProvider()));
 
@@ -192,7 +263,7 @@ public final class GrpcStorageOptions extends StorageOptions
         // this is totally valid. instead we want to monitor if the stream is doing work and if not
         // timeout.
         .setIdleTimeout(totalTimeout);
-    return builder.build();
+    return Tuple.of(builder.build(), defaultOpts);
   }
 
   /** @since 2.14.0 This new api is in preview and is subject to breaking changes. */
@@ -502,8 +573,11 @@ public final class GrpcStorageOptions extends StorageOptions
       if (options instanceof GrpcStorageOptions) {
         GrpcStorageOptions grpcStorageOptions = (GrpcStorageOptions) options;
         try {
-          StorageSettings storageSettings = grpcStorageOptions.getStorageSettings();
-          return new GrpcStorageImpl(grpcStorageOptions, StorageClient.create(storageSettings));
+          Tuple<StorageSettings, Opts<UserProject>> t = grpcStorageOptions.resolveSettingsAndOpts();
+          StorageSettings storageSettings = t.x();
+          Opts<UserProject> defaultOpts = t.y();
+          return new GrpcStorageImpl(
+              grpcStorageOptions, StorageClient.create(storageSettings), defaultOpts);
         } catch (IOException e) {
           throw new IllegalStateException(
               "Unable to instantiate gRPC com.google.cloud.storage.Storage client.", e);

--- a/google-cloud-storage/src/main/java/com/google/cloud/storage/UnifiedOpts.java
+++ b/google-cloud-storage/src/main/java/com/google/cloud/storage/UnifiedOpts.java
@@ -2288,6 +2288,22 @@ final class UnifiedOpts {
           .orElse(Decoder.identity());
     }
 
+    <L extends T> Opts<T> prepend(Opts<L> l) {
+      Opts<T> right = this;
+      Set<? extends Class<? extends Opt>> rightClasses =
+          right.opts.stream().map(Opt::getClass).collect(Collectors.toSet());
+
+      ImmutableList<T> list =
+          Stream.of(
+                  l.opts.stream()
+                      // remove opts which are defined on the right hand size
+                      .filter(o -> !rightClasses.contains(o.getClass())),
+                  right.opts.stream())
+              .flatMap(x -> x)
+              .collect(ImmutableList.toImmutableList());
+      return new Opts<>(list);
+    }
+
     private Mapper<ImmutableMap.Builder<StorageRpc.Option, Object>> rpcOptionMapper() {
       return fuseMappers(RpcOptVal.class, RpcOptVal::mapper);
     }
@@ -2308,6 +2324,10 @@ final class UnifiedOpts {
 
     static <T extends Opt> Opts<T> from(T... ts) {
       return new Opts<>(ImmutableList.copyOf(ts));
+    }
+
+    static <O extends Opt> Opts<O> empty() {
+      return new Opts<>(ImmutableList.of());
     }
 
     /**

--- a/google-cloud-storage/src/main/java/com/google/cloud/storage/UnifiedOpts.java
+++ b/google-cloud-storage/src/main/java/com/google/cloud/storage/UnifiedOpts.java
@@ -2288,17 +2288,22 @@ final class UnifiedOpts {
           .orElse(Decoder.identity());
     }
 
-    <L extends T> Opts<T> prepend(Opts<L> l) {
-      Opts<T> right = this;
-      Set<? extends Class<? extends Opt>> rightClasses =
-          right.opts.stream().map(Opt::getClass).collect(Collectors.toSet());
+    /**
+     * Create a new instance of Opts<T> where {@code toPrepend} and {@code this}. If an {@link Opt}
+     * type ({@code Class<T>}) is present in both {@code toPrepend} and {@code this}, the {@link
+     * Opt} from {@code this} will take priority when applied via one of the produced mappers.
+     */
+    Opts<T> prepend(Opts<? extends T> toPrepend) {
+      // inventory the Opt types already present in this
+      Set<? extends Class<? extends Opt>> existingOptTypes =
+          this.opts.stream().map(Opt::getClass).collect(Collectors.toSet());
 
       ImmutableList<T> list =
           Stream.of(
-                  l.opts.stream()
-                      // remove opts which are defined on the right hand size
-                      .filter(o -> !rightClasses.contains(o.getClass())),
-                  right.opts.stream())
+                  toPrepend.opts.stream()
+                      // exclude those opt types which are already present in this
+                      .filter(o -> !existingOptTypes.contains(o.getClass())),
+                  this.opts.stream())
               .flatMap(x -> x)
               .collect(ImmutableList.toImmutableList());
       return new Opts<>(list);

--- a/google-cloud-storage/src/test/java/com/google/cloud/storage/TransportCompatibilityTest.java
+++ b/google-cloud-storage/src/test/java/com/google/cloud/storage/TransportCompatibilityTest.java
@@ -22,6 +22,7 @@ import com.google.cloud.NoCredentials;
 import com.google.cloud.storage.PostPolicyV4.PostConditionsV4;
 import com.google.cloud.storage.PostPolicyV4.PostFieldsV4;
 import com.google.cloud.storage.TransportCompatibility.Transport;
+import com.google.cloud.storage.UnifiedOpts.Opts;
 import com.google.common.collect.ImmutableList;
 import java.util.function.Supplier;
 import java.util.stream.Stream;
@@ -37,7 +38,7 @@ public final class TransportCompatibilityTest {
             .setCredentials(NoCredentials.getInstance())
             .build();
     @SuppressWarnings("resource")
-    Storage s = new GrpcStorageImpl(options, null);
+    Storage s = new GrpcStorageImpl(options, null, Opts.empty());
     ImmutableList<String> messages =
         Stream.<Supplier<?>>of(
                 s::batch,

--- a/google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITQuotaProjectIdTest.java
+++ b/google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITQuotaProjectIdTest.java
@@ -1,0 +1,159 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.storage.it;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.junit.Assume.assumeTrue;
+
+import com.google.api.gax.paging.Page;
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BucketInfo;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.Storage.BlobListOption;
+import com.google.cloud.storage.StorageOptions;
+import com.google.cloud.storage.TransportCompatibility.Transport;
+import com.google.cloud.storage.it.runner.StorageITRunner;
+import com.google.cloud.storage.it.runner.annotations.Backend;
+import com.google.cloud.storage.it.runner.annotations.BucketFixture;
+import com.google.cloud.storage.it.runner.annotations.BucketType;
+import com.google.cloud.storage.it.runner.annotations.CrossRun;
+import com.google.cloud.storage.it.runner.annotations.Inject;
+import com.google.cloud.storage.it.runner.registry.ObjectsFixture;
+import java.util.stream.StreamSupport;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+@RunWith(StorageITRunner.class)
+@CrossRun(
+    backends = Backend.PROD,
+    transports = {Transport.HTTP, Transport.GRPC})
+public final class ITQuotaProjectIdTest {
+  // a project number to a dev project which is not used for CI
+  // this value is the negative case, and points to a project which doesn't have access to the
+  // bucket used here.
+  private static final String BAD_PROJECT_ID = "954355001984";
+
+  @Inject public Storage storage;
+
+  @Inject
+  @BucketFixture(BucketType.REQUESTER_PAYS)
+  public BucketInfo bucket;
+
+  // make sure there is an object in the bucket to be listed
+  @Inject
+  @BucketFixture(BucketType.REQUESTER_PAYS)
+  public ObjectsFixture objectsFixture;
+
+  private StorageOptions baseOptions;
+  private String projectId;
+  private GoogleCredentials credentials;
+
+  @Before
+  public void setUp() throws Exception {
+    baseOptions = storage.getOptions();
+    assumeTrue(
+        "These tests require GoogleCredentials",
+        baseOptions.getCredentials() instanceof GoogleCredentials);
+    credentials = (GoogleCredentials) baseOptions.getCredentials();
+    projectId = baseOptions.getProjectId();
+  }
+
+  /*
+   * UserProject precedence
+   * 1. Method userProject Option
+   * 2. ServiceOptions.getQuotaProjectId()
+   * 3. Credentials.quota_project_id
+   */
+
+  @Test
+  public void fromCredentials() throws Exception {
+    StorageOptions build =
+        baseOptions
+            .toBuilder()
+            .setCredentials(credentialsWithQuotaProjectId(credentials, projectId))
+            .build();
+
+    try (Storage s = build.getService()) {
+      Page<Blob> page = s.list(bucket.getName());
+      assertPage(page);
+    }
+  }
+
+  @Test
+  public void methodOptionOverCredentials() throws Exception {
+    StorageOptions build =
+        baseOptions
+            .toBuilder()
+            .setCredentials(credentialsWithQuotaProjectId(credentials, BAD_PROJECT_ID))
+            .build();
+
+    try (Storage s = build.getService()) {
+      Page<Blob> page = s.list(bucket.getName(), BlobListOption.userProject(projectId));
+      assertPage(page);
+    }
+  }
+
+  @Test
+  public void fromServiceOptionParameter() throws Exception {
+    StorageOptions build = baseOptions.toBuilder().setQuotaProjectId(projectId).build();
+
+    try (Storage s = build.getService()) {
+      Page<Blob> page = s.list(bucket.getName());
+      assertPage(page);
+    }
+  }
+
+  @Test
+  public void serviceOptionParameterOverCredentials() throws Exception {
+    StorageOptions build =
+        baseOptions
+            .toBuilder()
+            .setCredentials(credentialsWithQuotaProjectId(credentials, BAD_PROJECT_ID))
+            .setQuotaProjectId(projectId)
+            .build();
+
+    try (Storage s = build.getService()) {
+      Page<Blob> page = s.list(bucket.getName());
+      assertPage(page);
+    }
+  }
+
+  @Test
+  public void methodOptionOverServiceOptionParameter() throws Exception {
+    StorageOptions build = baseOptions.toBuilder().setQuotaProjectId(BAD_PROJECT_ID).build();
+
+    try (Storage s = build.getService()) {
+      Page<Blob> page = s.list(bucket.getName(), BlobListOption.userProject(projectId));
+      assertPage(page);
+    }
+  }
+
+  private void assertPage(Page<Blob> page) {
+    boolean info1InResults =
+        StreamSupport.stream(page.iterateAll().spliterator(), false)
+            .map(Blob::getName)
+            .anyMatch(objectsFixture.getInfo1().getName()::equals);
+    assertThat(info1InResults).isTrue();
+  }
+
+  private GoogleCredentials credentialsWithQuotaProjectId(
+      GoogleCredentials creds, String quotaProjectId) {
+    return creds.toBuilder().setQuotaProjectId(quotaProjectId).build();
+  }
+}

